flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/utils/BulkFormatRecordReader.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.storage.filestore.utils;
+
+import org.apache.flink.connector.file.src.reader.BulkFormat;
+import org.apache.flink.connector.file.src.util.RecordAndPosition;
+import org.apache.flink.table.data.RowData;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+
+/** Wrap a {@link BulkFormat.Reader<RowData>} as a {@link RecordReader<T>}. */
+public class BulkFormatRecordReader<T> implements RecordReader<T> {
+
+    private final ObjectSerializer<T> serializer;
+    private final BulkFormat.Reader<RowData> reader;
+
+    public BulkFormatRecordReader(
+            ObjectSerializer<T> serializer, BulkFormat.Reader<RowData> reader) {
+        this.serializer = serializer;
+        this.reader = reader;
+    }
+
+    @Nullable
+    @Override
+    public RecordIterator<T> readBatch() throws IOException {
+        BulkFormat.RecordIterator<RowData> iterator = reader.readBatch();
+        return iterator == null ? null : new BulkFormatIteratorWrapper(iterator);
+    }
+
+    @Override
+    public void close() throws IOException {
+        reader.close();
+    }
+
+    private class BulkFormatIteratorWrapper implements RecordIterator<T> {
+
+        private final BulkFormat.RecordIterator<RowData> iterator;
+        private T current;
+
+        private BulkFormatIteratorWrapper(BulkFormat.RecordIterator<RowData> iterator) {
+            this.iterator = iterator;
+            this.current = null;
+        }
+
+        @Override
+        public boolean advanceNext() throws IOException {
+            RecordAndPosition<RowData> next = iterator.next();
+            if (next == null) {
+                return false;
+            }
+            current = serializer.fromRow(next.getRecord());
+            return true;
+        }
+
+        @Override
+        public boolean singleInstance() {
+            return true;
```

Comment:
Can we implement a version with singleInstance false?

Reply:
Not for now. Let's implement it when it is really needed.

Reply:
What is the purpose of this PR? What is using reader?